File: app/src/main/java/com/example/savio/testeandroid/view/FormFragment.java
package com.example.savio.testeandroid.view;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.util.Patterns;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;

import com.example.savio.testeandroid.R;
import com.example.savio.testeandroid.SuccessActivity;
import com.example.savio.testeandroid.contract.FormContract;
import com.example.savio.testeandroid.mask.PhoneMask;
import com.example.savio.testeandroid.presenter.FormPresenter;

import java.lang.ref.WeakReference;
import java.util.List;

public class FormFragment extends Fragment implements FormContract.View{

    private FormPresenter formPresenter;

    PhoneMask phoneMask;

    View view;

    TextView title;
    EditText name, email, phone;
    CheckBox check;
    Button btnSend;

    public FormFragment() {}


//-----------------------------------LifeCycle------------------------------------------------------//

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_form, container, false);

        //instance the presenter
        formPresenter = new FormPresenter(this);

        //call the function to get de datas
        formPresenter.requestData();

        return view;

    }//onCreate ends


//-----------------------------------Override-------------------------------------------------------//

    @Override
    public void initView() {

        title = view.findViewById(R.id.form_title);
        name = view.findViewById(R.id.form_name);
        email = view.findViewById(R.id.form_email);
        phone = view.findViewById(R.id.form_phone);
        check = view.findViewById(R.id.form_check);
        btnSend =  view.findViewById(R.id.form_btn);

        //add the mask of the phone number
        phoneMask = new PhoneMask(new WeakReference<>(phone));
        phone.addTextChangedListener(phoneMask);

        //wait the button click
        btnSend.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                fieldValidation();
            }
        });
    }

    @Override
    public void setViewData(List<Object> messages) {

        title.setText(messages.get(0).toString());
        name.setHint(messages.get(1).toString());
        email.setHint(messages.get(2).toString());
        phone.setHint(messages.get(3).toString());
        check.setText(messages.get(4).toString());
        btnSend.setText(messages.get(5).toString());
    }


//-----------------------------------My Functions---------------------------------------------------//

    //resume all valitions of the fields
    public void fieldValidation(){

        //empty validation
        if(!TextUtils.isEmpty(name.getText()) && !TextUtils.isEmpty(email.getText()) && !TextUtils.isEmpty(phone.getText())){

            //email address validation
            if(Patterns.EMAIL_ADDRESS.matcher(email.getText()).matches() ){

                if(phone.getText().toString().length() > 13){

                    //open successful screen
                    startActivity(new Intent(getContext(), SuccessActivity.class));
                }
                else{

                    validationMessage("Não é um telefone válido.");
                }
            }
            else{

                validationMessage("Não é um email válido.");
            }
        }
        else{

            validationMessage("Todos os campos são obrigatórios.");
        }
    }

    //show tooltip message
    public void validationMessage(String message){

        Snackbar snackbar = Snackbar.make(view.findViewById(R.id.layout_form), message, Snackbar.LENGTH_LONG);
        snackbar.show();
    }

}

File: app/src/main/java/com/rafhack/testeandroid/form/FormFragment.kt
package com.rafhack.testeandroid.form

import android.support.constraint.ConstraintLayout
import android.support.design.widget.Snackbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import com.rafhack.testeandroid.R
import com.rafhack.testeandroid.base.BaseProgressFragment
import com.rafhack.testeandroid.data.entities.Cell
import com.rafhack.testeandroid.data.entities.Type

class FormFragment : BaseProgressFragment(), FormContract.View {

    private lateinit var linContainer: LinearLayout
    private lateinit var ctlSuccess: ConstraintLayout
    private lateinit var tvwSendNew: TextView
    private var manager: DynamicFormManager? = null
    private val presenter = FormPresenter(this)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?): View {
        val view = inflater.inflate(R.layout.fragment_form, container, true)
        linContainer = view.findViewById(R.id.fragment_form_lin_content)
        ctlSuccess = view.findViewById(R.id.fragment_form_ctl_success)
        tvwSendNew = view.findViewById(R.id.fragment_form_tvw_send_new)
        setupListner()
        presenter.getCells()
        return view
    }

    private fun setupListner() {
        tvwSendNew.setOnClickListener {
            manager?.clearForm()
            linContainer.visibility = View.VISIBLE
            ctlSuccess.visibility = View.GONE
        }
    }

    private val validatedFieldsListener: ((type: Type) -> Unit) = {
        if (it == Type.SEND) {
            linContainer.visibility = View.GONE
            ctlSuccess.visibility = View.VISIBLE
        }
    }

    override fun setProgress(active: Boolean) {
        if (active) showProgress() else hideProgress()
    }

    override fun showErrorMessage(message: String) {
        Snackbar.make(view?.findViewById(R.id.fragment_base_progress_ctl_root)!!,
                message, Snackbar.LENGTH_LONG).show()
    }

    override fun inflateCells(cells: ArrayList<Cell>) {
        manager = DynamicFormManager(linContainer, validatedFieldsListener)
        manager?.cells = cells
    }

}